dbms/src/Processors/Formats/Impl/ParquetBlockInputFormat.cpp

```diff
@@ -4,91 +4,171 @@
+        return arrow::Status::OK();
+    }
 
-            if (row_group_current < row_group_total)
-                throw Exception{"Got new data, but data from previous chunks was not read " +
-                                std::to_string(row_group_current) + "/" + std::to_string(row_group_total),
-                                ErrorCodes::CANNOT_READ_ALL_DATA};
+    virtual arrow::Status Tell(int64_t* position) const override
+    {
+        *position = in.getPosition();
+        return arrow::Status::OK();
+    }
 
-            file_data.clear();
-            {
-                WriteBufferFromString file_buffer(file_data);
-                copyData(in, file_buffer);
-            }
-
-            buffer = std::make_unique<arrow::Buffer>(file_data);
-            // TODO: maybe use parquet::RandomAccessSource?
-            auto status = parquet::arrow::FileReader::Make(
-                    ::arrow::default_memory_pool(),
-                    parquet::ParquetFileReader::Open(std::make_shared<::arrow::io::BufferReader>(*buffer)),
-                    &file_reader);
-
-            row_group_total = file_reader->num_row_groups();
-            row_group_current = 0;
-        }
-        //DUMP(row_group_current, row_group_total);
-        if (row_group_current >= row_group_total)
-            return res;
+    virtual bool closed() const override { return is_closed; }
 
-        // TODO: also catch a ParquetException thrown by filereader?
-        //arrow::Status read_status = filereader.ReadTable(&table);
-        std::shared_ptr<arrow::Table> table;
-        arrow::Status read_status = file_reader->ReadRowGroup(row_group_current, &table);
+    virtual arrow::Status Read(int64_t nbytes, int64_t* bytes_read, void* out) override
+    {
+        *bytes_read = in.readBig(reinterpret_cast<char *>(out), nbytes);
+        return arrow::Status::OK();
+    }
 
-        ArrowColumnToCHColumn::arrowTableToCHChunk(res, table, read_status, header, row_group_current, "Parquet");
-        return res;
+    virtual arrow::Status Read(int64_t nbytes, std::shared_ptr<arrow::Buffer>* out) override
+    {
+        std::shared_ptr<arrow::Buffer> buf;
+        ARROW_RETURN_NOT_OK(arrow::AllocateBuffer(nbytes, &buf));
+        size_t n = in.readBig(reinterpret_cast<char *>(buf->mutable_data()), nbytes);
+        *out = arrow::SliceBuffer(buf, 0, n);
+        return arrow::Status::OK();
     }
 
-    void ParquetBlockInputFormat::resetParser()
+    virtual arrow::Status Seek(int64_t position) override
     {
-        IInputFormat::resetParser();
+        in.seek(position, SEEK_SET);
+        return arrow::Status::OK();
+    }
+
+private:
+    SeekableReadBuffer& in;
+    off_t file_size;
+    bool is_closed;
+};
+
+
+static std::shared_ptr<arrow::io::RandomAccessFile>  as_arrow_file(ReadBuffer & in)
+{
+    if (auto fd_in = dynamic_cast<ReadBufferFromFileDescriptor*>(&in))
+    {
+        struct stat stat;
+        ::fstat(fd_in->getFD(), &stat);
```

Comment:
It's kinda hacky due to limitations of ReadBuffer interface.

I think there is a need for another abstraction to support random access and splittable files while supporting different filesystems e.g. local/s3/hdfs.
e.g.
```
class RandomAccessFile 
{
  off_t file_size();
  ReadBuffer readAt(off_t offset);
  ReadBuffer readAt(off_t offset, size_t size);
}
```

Random access use cases:
Parquet/Orc: get file size to find file metadata in footer, then read multiple row groups parallel
Avro: read header, then split at random offsets to read in parallel (avro has synchronization markers)
Any newline separated format: split at  random offsets, sync on newline, read in parallel.

That would be especially great with HDSFS/S3.

Reply:
We already have this interface, it's named `SeekableReadBuffer`

Reply:
(dynamic_cast is ok).

Reply:
I see that you're already using it in exactly this way :)